superset/db_engine_specs/databricks.py

```diff
@@ -434,7 +434,25 @@ def get_default_catalog(
         cls,
         database: Database,
     ) -> str | None:
+        """
+        Return the default catalog.
+
+        The default behavior for Databricks is confusing. When Unity Catalog is not
+        enabled we have (the DB engine spec hasn't been tested with it enabled):
+
+            > SHOW CATALOGS;
+            spark_catalog
+            > SELECT current_catalog();
+            hive_metastore
+
+        To handle permissions correctly we use the result of `SHOW CATALOGS` when a
+        single catalog is returned.
+        """
         with database.get_sqla_engine() as engine:
+            catalogs = {catalog for (catalog,) in engine.execute("SHOW CATALOGS")}
+            if len(catalogs) == 1:
+                return catalogs.pop()
```

Comment:
if `SHOW CATALOGS` returns only `spark_catalog`, it seems we're returning it. Do we want to return `spark_catalog`, or do we need to return `hive_metastore` instead (when there's only `spark_catalog`)? I'm asking it because I think there are issues with other metadata calls if we try to use `spark_catalog` when the catalog selected in the form is `hive_metastore` (or if UC is not enabled).

Also, if `multi_catalog` is disabled, do we want to rely on `SHOW CATALOGS`? Should we validate if `multi_catalog` is disabled first, and if so return the value of `current_catalog()` directly (which should be the catalog from the form)?

Reply:
The problem is that we need to create the permissions regardless of the state of `multi_catalog`. For Databricks, the migration process will create permissions for:

1. Whatever `get_default_catalog()` returns.
2. All the catalogs from `get_catalog_names` different from the default catalog from (1).

For Databricks, this means we're creating 2 permissions when Unity Catalog is not enabled:

1. `[Databricks].[hive_metastore]`
2. `[Databricks].[spark_catalog]`

This in itself is already a problem for the admin — which would should they assign roles to? Either one is problematic, depending on the status of `multi_catalog`:

With `multi_catalog` disabled the catalog sent by API calls is `null`, so it gets replaced with the default catalog, `hive_metastore`. If the admin assign roles to `[Databricks].[spark_catalog]` people won't be able to access datasets, because of the name mismatch.

With `multi_catalog` enabled the only option in the dropdown is `spark_catalog`. Which means if the admin assigned roles to `[Databricks].[hive_metastore]` people won't be able to access datasets.

With the change in this PR, everything is consistent, because the default catalog is part of the list of catalogs.